lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -0,0 +1,580 @@
+      }
+      for (auto n : llvm::reverse(currentPath)) {
+        FieldRef f = getFieldRefFromValue(n);
+        auto signalName = getFieldName(f).first;
+        if (!signalName.empty() && lastSignalName != signalName)
+          errorDiag << module.getName() << "." << signalName << " <- ";
+        lastSignalName = signalName;
+        if (n == node)
+          break;
+      }
+      currentPath.push_back(node);
+      return failure();
+    }
+    auto tmpVisiting = visiting;
+    visiting.insert(node);
+    SmallVector<Value> valuesForSameStorageLoc = {node};
+
+    SmallVector<ValueOpPair> children;
+    // Now get all the Values, that refer to the same storage location as
+    // `node`.
+    auto storageLocIter = valToStorageLocIdMap.find(node);
+    if (storageLocIter != valToStorageLocIdMap.end())
+      // A Value can refer to multiple Storage locations, (for example, result
+      // of subaccess op).
+      for (auto storageLocId : storageLocIter->getSecond())
+        // All the values that refer to `storageLocId` are a valid child of
+        // `node`.
+        for (auto aliasVal : storageLocIdToValuesMap[storageLocId])
+          // The following check ensures, Values referring to the same Location
+          // Id are added only once.
+          if (aliasVal != node) {
+            valuesForSameStorageLoc.push_back(aliasVal);
+            visiting.insert(aliasVal);
+          }
+    currentPath.push_back(node);
+    for (auto visitingNode : valuesForSameStorageLoc) {
+      // Record the current value being traversed, used for printing the
+      // discovered cycle.
+      currentPath.push_back(visitingNode);
+      // Get the Values that have an edge from this node. The ValueOpPair
+      // records the Operation that is the owner of the corresponding child
+      // value. The operation will only be used for error reporting.
+      getChildren(visitingNode, inputArg, children);
+      // Now continue the DFS recursively.
+      for (auto childNode : children)
+        if (dfsFromNode(childNode.first, inputArg).failed()) {
+          printError(visitingNode, childNode);
+          return failure();
+        }
+      currentPath.pop_back();
+    }
+    currentPath.pop_back();
+
+    visiting = std::move(tmpVisiting);
+    // Finished discovering all the reachable nodes from node.
+    visited.insert(node);
+    return success();
+  }
+
+  // Local implementation of getFieldRefFromValue, returns the Aggregate type
+  /// Get the FieldRef from a value. Customized implementation of
+  /// getFieldRefFromValue. If `field` is a subfield, then this will return the
+  /// base aggregate Value and the corresponding fieldId. This will return
+  /// nullptr if thie field is a result of a register and can be ignored.
+  FieldRef getFieldRef(Value field) {
+    if (field.isa<BlockArgument>() ||
+        !isa<SubfieldOp, SubaccessOp, SubindexOp>(field.getDefiningOp()))
```

Comment:
I think this is simpler:
```suggestion
    if (!isa_and_nonnull<SubfieldOp, SubaccessOp, SubindexOp>(field.getDefiningOp()))
```

Reply:
Thanks, updated this.